lib/tests/test_pathutils.c

```diff
@@ -88,3 +106,47 @@ Test(test_pathutils, test_get_filename_extension)
   cr_assert_null(get_filename_extension(""), "wrong file name extension returned");
   cr_assert_null(get_filename_extension(NULL), "wrong file name extension returned");
 }
+
+void
+test_get_filenames_setup(void)
+{
+  for (gint i = 0; i < dirs_len; ++i)
+    cr_assert_not(mkdir(dirs[i], S_IRUSR | S_IWUSR | S_IXUSR) < 0, "mkdir error %s", dirs[i]);
+
+  for (gint i = 0; i < files_len; ++i)
+    {
+      int fd = open(files[i], O_CREAT | O_RDWR, 0644);
+      cr_assert_not(fd < 0, "open error %s", files[i]);
```

Comment:
Thanks for the heads up, I did not check that. :)

After a digging a bit, I believe we can assert there.

According to https://criterion.readthedocs.io/en/master/starter.html
"If a setup crashes, you will get a warning message, and the test will be aborted and marked as a failure. Is a teardown crashes, you will get a warning message, and the test will keep its result."

I added `cr_assert(FALSE);` as the first line of my setup function, and got the following:
```
[14:31] ~/Work/build/OSE $ lib/tests/test_pathutils
[----] ../../repos/OSE/lib/tests/test_pathutils.c:113: Assertion failed: The expression (0) is false.
[----] Warning! The test `test_pathutils::test_get_sorted_filenames` crashed during its setup or teardown.
[====] Synthesis: Tested: 6 | Passing: 5 | Failing: 1 | Crashing: 1 
```

With these, I believe asserting is allowed in setup and teardown. What do you think?

Reply:
A crash is different than an assertion. AFAIK assertions inside setup and teardown are not documented. You can ask Snaipe to document the behavior, but until then, I would not use assert in setup().

Reply:
Well, adding an assertion to `setup()` resulted in a SIGSEGV in a longjmp call inside the forked test case, which is something unexpected IMHO (it might be a bug in Criterion, or just ASLR again).

I'll investigate this. `g_assert()` is safer to use here in my opinion.

Reply:
Does asserting if the mkdir, or create file was successful? I mean with that you are testing if mkdir or open works.
It is possible disk has issue, but debugging it is going to be nasty either way.
There is always the alternative, to create those files and store them in git, you can be sure they exists and no setup or assertion is needed.
This is not ideal either, as the paint point is that the tested code cannot function without disk.

Imho if you have to assert in setup phase, that's probably issue in your test strategy, as one test should test one thing.

Reply:
Does asserting if the mkdir, or create file was successful? I mean with that you are testing if mkdir or open works.
It is possible disk has issue, but debugging it is going to be nasty either way.
There is always the alternative, to create those files and store them in git, you can be sure they exists and no setup or assertion is needed.
This is not ideal either, as the paint point is that the tested code cannot function without disk.

Imho if you have to assert in setup phase, that's probably issue in your test strategy, as one test should test one thing.

Reply:
Personally I'm on the assert side, because I think it helps the debugging if we report the problem at the point of origin. Can we discuss this IRL?

Reply:
sure.

Reply:
I changed it to `g_assert()`.

Reply:
Thanks.

Reply:
In general, I agree with @Kokan. We should not assert in setup and teardown functions. They should contain initialization code that is not part of the actual test case, containing unconditional truths. If the setup/teardown logic contains something that can fail, a separate test case should be added to test that logic, and every other test case should assume that the initialization logic is correct and won't fail anymore.

This is true for unit tests, where ideally all unpredictable resources are mocked out.
For example, you could have mocked `g_dir_read_name` and other disk operations out from `pdb_get_filenames`, so you could write a unit test for the code you wrote, not for the code you wrote plus glib's disk functions and all the other unpredictable factors that a file system operation might cause.

JUnit has methods for "stating assumptions about the conditions in which a test is meaningful".
https://junit.org/junit4/javadoc/4.12/org/junit/Assume.html
This would be perfect for your situation.
You test case is about 2 things: testing `pdb_get_filenames` and `pdb_sort_filenames`.
The test about `pdb_sort_filenames` is "not meaningful" if you can't fetch its input, but failing to fetch filenames does not mean that `pdb_sort_filenames` is broken.

I'll keep all this in mind when rewriting our worst-looking test cases.

In case you write a functional/component test, you might want to assert initialization code as well, but I would not do that in setup() and teardown(), especially because Criterion can only handle a _crash_ in those functions, not assertion failures (by design IMHO).

----

I don't want to request a change on this PR, I just wanted to share how I would like to write tests from now on. The JUnit guys have really good principles/paradigms for writing clean and beautiful tests, a few might apply to our low-level environment as well in my opinion.